### PR TITLE
Support node selector and tolerations

### PR DESCRIPTION
**Pull request by philipwu08**
https://github.com/stolostron/backlog/issues/26684

Signed-off-by: Philip Wu <phwu@redhat.com>

### CODE REVIEW

File: addon/addon.go [hunk 1]
```diff
@@ -149,20 +152,35 @@ func NewAddonManager(kubeConfig *rest.Config, agentImage string, agentInstallAll
...
+			schema.GroupVersionResource{Version: "v1", Resource: "configmaps"},
+			schema.GroupVersionResource{Group: "addon.open-cluster-management.io", Version: "v1alpha1", Resource: "addondeploymentconfigs"},
+		).
+		WithGetValuesFuncs(
+			// get the AddOnDeloymentConfig object and transform it to Values object
+			addonfactory.GetAddOnDeloymentConfigValues(
+				addonfactory.NewAddOnDeloymentConfigGetter(addonClient),
+				addonfactory.ToAddOnNodePlacementValues,
+			),
+			getValue,
```
Comment (skeeey) [inline addon/addon.go:178]: we still need `addonfactory.GetValuesFromAddonAnnotation`, we use addon value annotation to configure the proxy and image registry

And I suggest put the `addonfactory.GetAddOnDeloymentConfigValues` at the end of valueFunc slice, thus we can use `AddOnDeloymentConfig` to override other configs

File: addon/addon.go [hunk 2]
```diff
@@ -149,20 +152,35 @@ func NewAddonManager(kubeConfig *rest.Config, agentImage string, agentInstallAll
...
 	if err != nil {
 		klog.Errorf("unable to create kube client: %v", err)
 		return addonMgr, err
 	}
 
 	agentFactory := addonfactory.NewAgentAddonFactory(AppMgrAddonName, ChartFS, ChartDir).
-		WithGetValuesFuncs(getValue, addonfactory.GetValuesFromAddonAnnotation).
+		// register the supported configuration types
+		WithConfigGVRs(
+			schema.GroupVersionResource{Version: "v1", Resource: "configmaps"},
```
Comment (skeeey) [inline addon/addon.go:169]: Do we use `configmap` as the configuration for application addon? if we do not, we do not need to add the `configmaps`

File: pkg/utils/subscription.go [hunk 3]
```diff
@@ -1242,6 +1242,19 @@ func CreateClusterManagementAddon(clt client.Client) {
 				Description: "Processes events and other requests to managed resources.",
 				DisplayName: "Application Manager",
 			},
+			SupportedConfigs: []addonV1alpha1.ConfigMeta{
+				{
+					ConfigGroupResource: addonV1alpha1.ConfigGroupResource{
+						Group:    "addon.open-cluster-management.io",
+						Resource: "addondeploymentconfigs",
```
Comment (skeeey) [inline pkg/utils/subscription.go:1249]: We don't need add addondeploymentconfig here, we want user add this manfully when user want to enable the addondeploymentcofngis as the addon configuration

**Comment by philipwu08** [comment]
@skeeey Please review the updated PR. Thanks.

File: addon/addon.go [hunk 4]
```diff
@@ -149,20 +151,35 @@ func NewAddonManager(kubeConfig *rest.Config, agentImage string, agentInstallAll
...
 
 	agentFactory := addonfactory.NewAgentAddonFactory(AppMgrAddonName, ChartFS, ChartDir).
-		WithGetValuesFuncs(getValue, addonfactory.GetValuesFromAddonAnnotation).
+		// register the supported configuration types
+		WithConfigGVRs(
+			schema.GroupVersionResource{Group: "addon.open-cluster-management.io", Version: "v1alpha1", Resource: "addondeploymentconfigs"},
+		).
+		WithGetValuesFuncs(
+			// get the AddOnDeloymentConfig object and transform it to Values object
+			addonfactory.GetAddOnDeloymentConfigValues(
```
Comment (skeeey) [inline addon/addon.go:172]: I think we need put the `GetAddOnDeloymentConfigValues` at the end, like
```
WithGetValuesFuncs(
   getValue,
   addonfactory.GetValuesFromAddonAnnotation,
   addonfactory.GetAddOnDeloymentConfigValues(...)
)
```

thus we can use addondeloymentconfig to override other configs, in future, we may deprecate the addon value annotation, in that time, we need a way to replace it

thanks

**Pull request merged**